### PR TITLE
Module 3: Add tests

### PR DESCRIPTION
**Pull request by JoelWee**
Thanks for the reviews so far :) @Jonesey13

### CODE REVIEW

File: .env.test [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Flask server configuration.
+FLASK_APP=app
+FLASK_ENV=development
+
+# Change the following values for local development.
+TRELLO_API_KEY=key
+TRELLO_API_TOKEN=token
+TRELLO_BOARD_ID=id
+FLASK_SKIP_DOTENV=True
```
Comment (Jonesey13) [inline .env.test]: We don't need this last one as we're not running flask directly

File: app.py [hunk 2]
```diff
@@ -1,30 +1,15 @@
-from flask import Flask, render_template, request, redirect, url_for
+from flask import Flask
 
-from trello_client import Trello
+from routes import bp
```
Comment (Jonesey13) [inline app.py:3]: Is there a particular reason we are using a blueprint here?

File: requirements.txt [hunk 3]
```diff
@@ -15,18 +15,27 @@ Jinja2==2.11.2
 lazy-object-proxy==1.4.3
```
Comment (Jonesey13) [inline requirements.txt:15]: Still not convinced we need all these dependencies

File: routes.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+from flask import Blueprint, redirect, render_template, request, url_for
```
Comment (Jonesey13) [inline routes.py:1]: I will assume most of this has been untouched from Module 2

File: templates/index.html [hunk 5]
```diff
@@ -8,22 +8,66 @@ <h1 class="display-4">To-Do App</h1>
...
+      {% endfor %}
+    </ul>
+  </div>
+
+  <div class="col-md-4">
+    <h2>Done</h2>
+    {% if view_model.show_all_done_items %}
+    <ul class="list-group mb-4">
+      {% for item in view_model.done_items  %}
+      <li>{{ item.title }}<form id="{{ 'reset-' + item.title|string }}"
```
Comment (Jonesey13) [inline templates/index.html]: Indentation here is a little strange

File: tests/test_items_view_model.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+from ToDoItem import ItemStatus, ToDoItem
+from view_models.items import ItemsViewModel
+
+
+@pytest.fixture
+def items_view_model():
+    return ItemsViewModel(
+        [
+            ToDoItem("a", ItemStatus.TO_DO, "a", datetime.now()),
+            ToDoItem("b", ItemStatus.TO_DO, "b", datetime(2020, 1, 1)),
```
Comment (Jonesey13) [inline tests/test_items_view_model.py]: You might want to extract `datetime(2020, 1, 1)` as a named const variable.

File: tests_e2e/test_app.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+@pytest.fixture(scope="module")
+def driver():
+    with webdriver.Firefox() as driver:
+        yield driver
+
+
+@pytest.fixture(scope="module")
+def test_app():
+    load_dotenv(".env")
```
Comment (Jonesey13) [inline tests_e2e/test_app.py]: This clashes with `load_dotenv(".env.test")` as only the first invoked load_dotenv() invocation is used.

This means these tests work fine when run individually but not when run altogether.

As a result you should probably:
* Combine the 2 configs
* Document this in the README

**Review by Jonesey13 — APPROVED** [review]
Looks good to me!

There's a small problem of running all the tests at once but apart from this everything appears to be working fine. 👍

File: app.py (hunk 2 above)
Comment (JoelWee) [reply app.py:3]: So that I wouldn't have to inline my routes in the create_app() method

File: requirements.txt (hunk 3 above)
Comment (JoelWee) [reply requirements.txt:15]: I think they might have come from black / flake8 but I'm not too sure... I'll leave it for now and try to clean it up when migrating to poetry

File: routes.py (hunk 4 above)
Comment (JoelWee) [reply routes.py:1]: yep just moved the file

File: tests_e2e/test_app.py (hunk 7 above)
Comment (JoelWee) [reply tests_e2e/test_app.py]: Wouldn't it be bad practice to give the unit tests access to an actual trello board though? 

Is there a way to reload the env file?

**Pull request merged**